### ISSUE TITLE
SIGABRT in ibus-x11 causing Emacs to hang

### ISSUE THREAD

**Issue by polaris64** [ibus/ibus#2484]
Which distribution and version?: 
Arch Linux (Linux 6.2.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 03 Mar 2023 15:58:31 +0000 x86_64 GNU/Linux)

Which desktop environment and version?: 
qtile 0.22.1

Which session type?: 
X11

Which application and version?: 
Emacs 29.0.60

IBus version?: 
1.5.28-1

Issue description:
Recently typing has become "sluggish" with frequent pauses. Typing in Emacs sometimes causes inputs to get rearranged (e.g. typing "hello" enters something like "llohe" into the buffer).

After a while Emacs hangs because `ibus-x11` crashes with `SIGABRT`. From the system journal: -

```
Mar 06 11:29:46 palenque systemd-coredump[143078]: [🡕] Process 1397 (ibus-x11) of user 1000 dumped core.

                                                   Stack trace of thread 1397:
                                                   #0  0x00007f8b611bd8ec n/a (libc.so.6 + 0x878ec)
                                                   #1  0x00007f8b6116eea8 raise (libc.so.6 + 0x38ea8)
                                                   #2  0x00007f8b6115853d abort (libc.so.6 + 0x2253d)
                                                   #3  0x00007f8b6115929e n/a (libc.so.6 + 0x2329e)
                                                   #4  0x00007f8b611c7657 n/a (libc.so.6 + 0x91657)
                                                   #5  0x00007f8b611c9442 n/a (libc.so.6 + 0x93442)
                                                   #6  0x00007f8b611cbe63 __libc_free (libc.so.6 + 0x95e63)
                                                   #7  0x00007f8b6135f94e XFree (libX11.so.6 + 0x4294e)
                                                   #8  0x00005570719e4311 n/a (ibus-x11 + 0xe311)
                                                   #9  0x00005570719e619a n/a (ibus-x11 + 0x1019a)
                                                   #10 0x00007f8b61e8159e n/a (libgdk-3.so.0 + 0x8b59e)
                                                   #11 0x00007f8b61e27029 gdk_display_get_event (libgdk-3.so.0 + 0x31029)
                                                   #12 0x00007f8b61e81a68 n/a (libgdk-3.so.0 + 0x8ba68)
                                                   #13 0x00007f8b614b582b g_main_context_dispatch (libglib-2.0.so.0 + 0x5582b)
                                                   #14 0x00007f8b6150ccc9 n/a (libglib-2.0.so.0 + 0xaccc9)
                                                   #15 0x00007f8b614b4d8f g_main_loop_run (libglib-2.0.so.0 + 0x54d8f)
                                                   #16 0x00007f8b61f262b2 ibus_main (libibus-1.0.so.5 + 0x372b2)
                                                   #17 0x00005570719d9505 n/a (ibus-x11 + 0x3505)
                                                   #18 0x00007f8b61159790 n/a (libc.so.6 + 0x23790)
                                                   #19 0x00007f8b6115984a __libc_start_main (libc.so.6 + 0x2384a)
                                                   #20 0x00005570719d96f5 n/a (ibus-x11 + 0x36f5)

                                                   Stack trace of thread 1415:
                                                   #0  0x00007f8b612309df __poll (libc.so.6 + 0xfa9df)
                                                   #1  0x00007f8b6150cc2f n/a (libglib-2.0.so.0 + 0xacc2f)
                                                   #2  0x00007f8b614b4d8f g_main_loop_run (libglib-2.0.so.0 + 0x54d8f)
                                                   #3  0x00007f8b61072aec n/a (libgio-2.0.so.0 + 0x10aaec)
                                                   #4  0x00007f8b614e2db5 n/a (libglib-2.0.so.0 + 0x82db5)
                                                   #5  0x00007f8b611bbbb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f8b6123dd90 n/a (libc.so.6 + 0x107d90)

                                                   Stack trace of thread 1412:
                                                   #0  0x00007f8b612309df __poll (libc.so.6 + 0xfa9df)
                                                   #1  0x00007f8b6150cc2f n/a (libglib-2.0.so.0 + 0xacc2f)
                                                   #2  0x00007f8b614b40e2 g_main_context_iteration (libglib-2.0.so.0 + 0x540e2)
                                                   #3  0x00007f8b614b4132 n/a (libglib-2.0.so.0 + 0x54132)
                                                   #4  0x00007f8b614e2db5 n/a (libglib-2.0.so.0 + 0x82db5)
                                                   #5  0x00007f8b611bbbb5 n/a (libc.so.6 + 0x85bb5)
                                                   #6  0x00007f8b6123dd90 n/a (libc.so.6 + 0x107d90)
                                                   ELF object binary architecture: AMD x86-64
```

Debugging the core dump with `gdb` and then running `thread apply all bt` yields: -

```
Thread 3 (Thread 0x7f8b5cfff6c0 (LWP 1412)):
warning: Section `.reg-xstate/1412' in core file too small.
#0  0x00007f8b612309df in __GI___poll (fds=0x557071c51f00, nfds=2,timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007f8b6150cc2f in g_main_context_poll (priority=<optimized out>, n_fds=2, fds=0x557071c51f00, timeout=<optimized out>,context=0x557071c53eb0) at ../glib/glib/gmain.c:4553
#2  g_main_context_iterate.constprop.0 (context=0x557071c53eb0,block=1, dispatch=1, self=<optimized out>) at ../glib/glib/gmain.c:4243
#3  0x00007f8b614b40e2 in g_main_context_iteration (context=0x557071c53eb0, may_block=may_block@entry=1) at ../glib/glib/gmain.c:4313
#4  0x00007f8b614b4132 in glib_worker_main (data=<optimized out>) at ../glib/glib/gmain.c:6416
#5  0x00007f8b614e2db5 in g_thread_proxy (data=0x557071c47b00) at ../glib/glib/gthread.c:831
#6  0x00007f8b611bbbb5 in start_thread (arg=<optimized out>) at pthread_create.c:444
#7  0x00007f8b6123dd90 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 2 (Thread 0x7f8b57fff6c0 (LWP 1415)):
warning: Section `.reg-xstate/1415' in core file too small.
#0  0x00007f8b612309df in __GI___poll (fds=0x557071c646d0, nfds=3,timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007f8b6150cc2f in g_main_context_poll (priority=<optimized out>, n_fds=3, fds=0x557071c646d0, timeout=<optimized out>,context=0x557071c62bf0) at ../glib/glib/gmain.c:4553
#2  g_main_context_iterate.constprop.0 (context=0x557071c62bf0,block=1, dispatch=1, self=<optimized out>) at ../glib/glib/gmain.c:4243
#3  0x00007f8b614b4d8f in g_main_loop_run (loop=0x557071c62ce0) at ../glib/glib/gmain.c:4448
#4  0x00007f8b61072aec in gdbus_shared_thread_func (user_data=0x557071c62bc0) at ../glib/gio/gdbusprivate.c:284
#5  0x00007f8b614e2db5 in g_thread_proxy (data=0x557071c5c640) at ../glib/glib/gthread.c:831
#6  0x00007f8b611bbbb5 in start_thread (arg=<optimized out>) at pthread_create.c:444
#7  0x00007f8b6123dd90 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 1 (Thread 0x7f8b5fd45980 (LWP 1397)):
#0  __pthread_kill_implementation (threadid=<optimized out>,signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007f8b611bd953 in __pthread_kill_internal (signo=6,threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007f8b6116eea8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f8b6115853d in __GI_abort () at abort.c:79
#4  0x00007f8b6115929e in __libc_message (fmt=fmt@entry=0x7f8b612d077e "%s\n") at ../sysdeps/posix/libc_fatal.c:150
#5  0x00007f8b611c7657 in malloc_printerr (str=str@entry=0x7f8b612d3590 "double free or corruption (fasttop)") at malloc.c:5651
#6  0x00007f8b611c9442 in _int_free (av=0x7f8b6130eaa0 <main_arena>,p=0x557071d50da0, have_lock=have_lock@entry=0) at malloc.c:4525
#7  0x00007f8b611cbe63 in __GI___libc_free (mem=mem@entry=0x557071d50db0) at malloc.c:3367
#8  0x00007f8b6135f94e in XFree (data=data@entry=0x557071d50db0) at /usr/src/debug/libx11/libX11-1.8.4/src/XlibInt.c:1633
#9  0x00005570719e4311 in ProcessQueue (connect_id=4,ims=0x557071e82e50) at ../../util/IMdkit/i18nPtHdr.c:1762
#10 _Xi18nMessageHandler (ims=ims@entry=0x557071e82e50, connect_id=4,p=p@entry=0x557071db6ab0 ">", delete=delete@entry=0x7ffe74734f84) at ../../util/IMdkit/i18nPtHdr.c:1923
#11 0x00005570719e619a in WaitXIMProtocol (dpy=<optimized out>,win=<optimized out>, ev=<optimized out>, client_data=0x557071e82e50 " \t\237qpU") at ../../util/IMdkit/i18nX.c:524
#12 0x00007f8b61e8159e in _gdk_x11_display_queue_events (display=0x557071bfe0e0) at ../gtk/gdk/x11/gdkeventsource.c:337
#13 0x00007f8b61e27029 in gdk_display_get_event (display=0x557071bfe0e0) at ../gtk/gdk/gdkdisplay.c:442
#14 0x00007f8b61e81a68 in gdk_event_source_dispatch.lto_priv () at ../gtk/gdk/x11/gdkeventsource.c:354
#15 0x00007f8b614b582b in g_main_dispatch (context=0x557071c27d20) at ../glib/glib/gmain.c:3454
#16 g_main_context_dispatch (context=0x557071c27d20) at ../glib/glib/gmain.c:4172
#17 0x00007f8b6150ccc9 in g_main_context_iterate.constprop.0 (context=0x557071c27d20, block=1, dispatch=1, self=<optimized out>) at ../glib/glib/gmain.c:4248
#18 0x00007f8b614b4d8f in g_main_loop_run (loop=0x557071d975e0) at ../glib/glib/gmain.c:4448
#19 0x00007f8b61f262b2 in ibus_main () at /usr/src/debug/ibus/ibus-1.5.28/src/ibusshare.c:327
#20 0x00005570719d9505 in main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/ibus/ibus-1.5.28/client/x11/main.c:1416
```

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
Unsure

Do you see any errors when you run ibus-daemon with the verbose option?:
Unsure: will need to do further research

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
Unsure: will need to do further research

**Comment by fujiwarat** [comment]
Duplicate of #2480

**Comment by fujiwarat** [comment]
Unfortunately I can reproduce this with the fix of #2480

Seems thsi is caused because ibus/util/IMdkit does not support the async process.
```
0x00007f8dc5718b94 in __pthread_kill_implementation () from /lib64/libc.so.6
(gdb) where
#0  0x00007f8dc5718b94 in __pthread_kill_implementation () at /lib64/libc.so.6
#1  0x00007f8dc56c7aee in raise () at /lib64/libc.so.6
#2  0x00007f8dc56b087f in abort () at /lib64/libc.so.6
#3  0x00007f8dc56b160f in _IO_peekc_locked.cold () at /lib64/libc.so.6
#4  0x00007f8dc5722ac5 in  () at /lib64/libc.so.6
#5  0x00007f8dc572492a in _int_free () at /lib64/libc.so.6
#6  0x00007f8dc57274ee in free () at /lib64/libc.so.6
#7  0x00007f8dc58b1ccd in XFree () at /lib64/libX11.so.6
#8  0x000055b5c90d6354 in ProcessQueue (ims=0x55b5c9ba54b0, connect_id=1)
    at ../../util/IMdkit/i18nPtHdr.c:1762
#9  0x000055b5c90d6778 in _Xi18nMessageHandler
    (ims=0x55b5c9ba54b0, connect_id=1, p=0x7f8da8005110 ">", delete=0x7ffc99166e04) at ../../util/IMdkit/i18nPtHdr.c:1923
#10 0x000055b5c90cc718 in WaitXIMProtocol
    (dpy=0x55b5c9b2a7b0, win=67108871, ev=0x7ffc99166e70, client_data=0x55b5c9ba54b0 "`\006\016ɵU") at ../../util/IMdkit/i18nX.c:524
#11 0x00007f8dc5b93741 in _gdk_x11_display_queue_events ()
    at /lib64/libgdk-3.so.0
#12 0x00007f8dc5b31adb in gdk_display_get_event () at /lib64/libgdk-3.so.0
#13 0x00007f8dc5b93c4e in gdk_event_source_dispatch.lto_priv ()
    at /lib64/libgdk-3.so.0
#14 0x00007f8dc5a0c7f8 in g_main_context_dispatch () at /lib64/libglib-2.0.so.0
#15 0x00007f8dc5a6af38 in g_main_context_iterate.isra ()
    at /lib64/libglib-2.0.so.0
#16 0x00007f8dc5a0beff in g_main_loop_run () at /lib64/libglib-2.0.so.0
#17 0x00007f8dc64ee616 in ibus_main ()
    at /root/rpmbuild/BUILD/ibus-1.5.28/src/ibusshare.c:327
#18 0x000055b5c90c7984 in main (argc=1, argv=0x7ffc991672a8)
    at /root/rpmbuild/BUILD/ibus-1.5.28/client/x11/main.c:1422
```

**Comment by fujiwarat** [comment]
Could you try to test my patch above?
